### DISCUSSION TITLE
Serialization of LocalDate scalar as parameter for GraphQLOperationRequest

### DISCUSSION THREAD

**Discussion by idegroot** [kobylynskyi/graphql-java-codegen#464] [Q&A] [answered]
Hi,

I hope this is the right place to ask my question.

I have a graphql shema that will give data based on a java.timeLocalDate like so:
```
scalar LocalDate
schema {
    query: Query
}
type QueryType {
    birthdays(dateOfBirth: LocalDate): [Person]!
}
```

The generated GraphQLOperationRequest implements the method with the LocalDate parameter which is great.

When the request get's serialized with request.toHttpJsonBody() the result is
```
{"query":"query { birthdays(dateOfBirth: 2020-12-23){ name } }"}
```

This query is rejected by the server because of the dashes in the String representation of the LocalDate.

When the value is quoted, the server accepts the request:
```
{"query":"query { birthdays(dateOfBirth: \"2020-12-23\"){ name } }"}
```

My quick solution is to change the parameter to type String and fill the parameter with localDate.toString(). Would there be another way to get the escaped quotes in the serialized request? I would prefer to use the LocalDate.

I am using the following plugin version to generate the client code:
```
<plugin>
    <groupId>io.github.kobylynskyi</groupId>
    <artifactId>graphql-codegen-maven-plugin</artifactId>
    <version>3.1.1</version>
</plugin>
```
Any advice will be very much appreciated!

Ida

**Comment by jxnu-liguobin** [comment]
As far as I know, double quotes are included only if the type is ```String```. 
You can try configuration```mappingConfig.setUseObjectMapperForRequestSerialization(singleton ("LocalDate")); ```? 
Of course, I'm not sure. 🤪 

gradle plugin
```
customTypesMapping = [
    ZonedDateTime: "java.time.ZonedDateTime"
]
useObjectMapperForRequestSerialization = [
    "ZonedDateTime"
]
```

**Reply by idegroot** [reply]
Thanks for your reply. Can you explain how and when I can set this property? Then I can give it a try.

**Reply by jxnu-liguobin** [reply]
I think you can take a look at this https://github.com/kobylynskyi/graphql-java-codegen/issues/260

**Reply by idegroot** [reply]
Thanks again for your answer, I really appreciate your help!

Issue #260 was indeed a similar case. However, the solution only works on a custom input type, not on the plain scalar type.

I define my ObjectMapper like this:
```
GraphQLRequestSerializer.OBJECT_MAPPER
               .registerModule(new JavaTimeModule())
               .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
```
And I define my Query like this:
```
type Query {
    birthdays(dateOfBirth: DateOfBirth!): [Person]!
}

input DateOfBirth {
    date: LocalDate!
}
```
I will get
```
{"query":"query { birthdays(dateOfBirth{date: \"2020-12-23\"}){ name } }"}
```

But when I define my Query like I want to:
```
type Query {
    birthdays(dateOfBirth: LocalDate): [Person]!
}
```
with the same objectmapper config I will still get
```
{"query":"query { birthdays(dateOfBirth: 2020-12-23){ name } }"}
```
I cannot get the solution to work on the GraphQLOperationRequest input fields directly.

**Reply by jxnu-liguobin** [reply]
Yes, because ```GraphQLRequestSerializer.getEntry(x, true)``` will only be available on custom input types. And scalars do not use ```ObjectMapper```.

**Reply by jxnu-liguobin** [reply]
You may need to use the mapping to String to solve this problem. Sorry, I can't help you any more. :octocat:

**Reply by idegroot** [reply]
Will do. Thanks for all your input!

**Comment by kobylynskyi** [comment]
@idegroot, in the description of your issue you mentioned that you are using 3.1.1
The solution described [here](https://github.com/kobylynskyi/graphql-java-codegen/issues/260#issuecomment-725199906) will work only starting from 4.0.0.
Could you please try upgrading to the latest version (as of today it is 4.0.1) and try again?
Thanks!

**Reply by idegroot** [reply]
@kobylynskyi, I did upgrade to 4.0.1 in my previous attempt. With this version I do get the objectmapper on my custom field with LocalDate but not when I put LocalDate as a parameter of the request directly like in the example.

**Reply by idegroot** [reply]
The full case is now:
dependency and plugin:
```
<dependency>
    <groupId>io.github.kobylynskyi</groupId>
    <artifactId>graphql-java-codegen</artifactId>
    <version>4.0.1</version>
</dependency>
<plugin>
    <groupId>io.github.kobylynskyi</groupId>
    <artifactId>graphql-codegen-maven-plugin</artifactId>
    <version>4.0.1</version>
    ....
```
Schema:
```
scalar LocalDate
schema {
    query: Query
}
type Query {
    birthdaysScalar(dateOfBirth: LocalDate!): [Person]!
    birthdaysType(dateOfBirth: DateOfBirth!): [Person]!
}
input DateOfBirth {
    date: LocalDate!
}
type Person {
    name: String!
    dateOfBirth: LocalDate!
}
```
Code for the request:
```
GraphQLRequestSerializer.OBJECT_MAPPER
        .registerModule(new JavaTimeModule())
        .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

PersonResponseProjection projection = new PersonResponseProjection().name();

BirthdaysScalarQueryRequest birthdaysScalarQueryRequest = BirthdaysScalarQueryRequest.builder()
        .setDateOfBirth(LocalDate.now())
        .build();
BirthdaysTypeQueryRequest birthdaysTypeQueryRequest = BirthdaysTypeQueryRequest.builder()
        .setDateOfBirth(DateOfBirth.builder().setDate(LocalDate.now()).build())
        .build();

GraphQLRequest requestScalar = new GraphQLRequest(birthdaysScalarQueryRequest, projection);
GraphQLRequest requestType = new GraphQLRequest(birthdaysTypeQueryRequest, projection);

System.out.println("Using Scalar: " + requestScalar.toHttpJsonBody());
System.out.println("Using Type: " + requestType.toHttpJsonBody());
```
Output:
```
Using Scalar: {"query":"query { birthdaysScalar: birthdaysScalar(dateOfBirth: 2020-12-29){ name } }"}
Using Type: {"query":"query { birthdaysType: birthdaysType(dateOfBirth: { date: \"2020-12-29\" }){ name } }"}
```

**Reply by kobylynskyi** [reply]
@idegroot, that's indeed not covered by the graphql-java-codegen. I will take a look at what can be done. Please stay tuned.

**Comment by jxnu-liguobin** [comment]
This involves how to properly serialize a Scalars if it is an imported class that cannot be modified. The #260 's unit test did not give the this test case(Scalars are used directly, not in ```type```),  only converted it to String type.May need a PR?

**Comment by kobylynskyi** [comment, marked as answer]
@idegroot, I've created a ticket for supporting your scenario (custom scalar as a parameter in GraphQL operation): #472

**Reply by kobylynskyi** [reply]
Changes were done in #473
Will release it as part of [4.1.0](https://github.com/kobylynskyi/graphql-java-codegen/milestone/27) this week (maybe even this year 😉)

**Reply by jxnu-liguobin** [reply]
Kotlin and Scala are supported in the latest pr. https://github.com/kobylynskyi/graphql-java-codegen/pull/474/files

**Reply by idegroot** [reply]
Wow, this is great. I downloaded the branch and tried it and can confirm it works for me. The output is now:
```
Using Scalar: {"query":"query { birthdaysScalar: birthdaysScalar(dateOfBirth: \"2020-12-30\"){ name } }"}
Using Type: {"query":"query { birthdaysType: birthdaysType(dateOfBirth: { date: \"2020-12-30\" }){ name } }"}
```
Thank you so much for your response and help. I am looking forward to the release of 4.1.0!

**Reply by kobylynskyi** [reply]
@idegroot, version [4.1.0](https://github.com/kobylynskyi/graphql-java-codegen/releases/tag/v4.1.0) with this fix was released today.

**Reply by idegroot** [reply]
It works like a charm, thank you both for fixing this so quickly!
